[Biomedical DC] Update answer message and answer table tile

juliawu opened the pull request:
This PR:
* Adds a copy button to the answer message tile
* Adds a download button to the answer table tile
* Slight tweaks to CSS styling to better match mocks
* Adds server logic for NL search results page to use biomedical-specific CSS if on biomedical.datacommons.org

Left for future PRs:
* Updated scrollbar styling: Some basic scrollbar styling has been applied, but getting it to match the mocks exactly is not straightforward and will take more development time. **Edit: No longer using scrollbars in table**
* Entity overview tile: This is a larger change, checking in the answer tiles first in the spirit of letting the team play with things earlier.

Screenshots for biomedical DC:

![Screenshot 2024-04-25 at 8 33 21 PM](https://github.com/datacommonsorg/website/assets/4034366/bfe65c51-b7da-48fe-8ee7-f3737c8ad86e)


![Screenshot 2024-04-25 at 8 34 51 PM](https://github.com/datacommonsorg/website/assets/4034366/d847025d-f0ce-4d3a-94f1-df8d1a0c09f7)


Screenshots for main DC:
![Screenshot 2024-04-25 at 8 36 59 PM](https://github.com/datacommonsorg/website/assets/4034366/a47c1aad-e635-4b4e-bf8f-3716b5ed06d2)


![Screenshot 2024-04-25 at 8 38 28 PM](https://github.com/datacommonsorg/website/assets/4034366/aae96a37-4c92-4d5d-9f35-bfba11dd17e0)


Example download csv:
Original query: "How many clinical variants associated with FGFR1 have clinical pathogenicity?:
Downloaded filename: `The assicated genetic variants for FGFR1, FGFR1 and 5 more other are.csv`
```
"DCID","Name","<-geneSymbol{typeOf:GeneticVariantGeneAssociation}->referenceSNPClusterID","<-geneSymbol{typeOf:GeneticVariantGeneAssociation}->referenceSNPClusterID source"
"bio/FGFR1","FGFR1","",""
"bio/galGal5_FGFR1","FGFR1","",""
"bio/galGal6_FGFR1","FGFR1","",""
"bio/hg19_FGFR1","FGFR1","bio/rs117756469; rs11996010; rs7823851; rs7829871; rs13274384; rs13248612; rs72636191; rs7845393; rs6996321; rs2003265; rs4739558; rs2010445; rs3758101; rs11992272; rs13261164; rs35699533; rs4236755; rs16887681; rs10958704; rs62505473; rs72630687; rs7815496; rs4733980; rs145530646; rs72634742; rs111690554; rs7824271; rs7001431; rs328291; rs4342555; rs66504003; rs13279569; rs17182127; rs328294; rs141893102; rs142164645; rs930828; rs2445024; rs146505289; rs12114071; rs2568232; rs4601278; rs1541897; rs17175624; rs12677818; rs4647906; rs117787947; rs2568235; rs35305468; rs56307815; rs4292654; rs17182002; rs328300; rs881299; rs1447055; bio/rs35045158; rs4733979; rs17176009; rs4733946; rs73675815; rs200246954; rs113932779; rs3213849; rs28618161; rs7825758; rs4082204; rs4739560; rs72632529; rs113991129; rs28877647; rs36061954; rs183333881; rs17182134; rs2445020; rs2468964; rs6990295; rs79295570; rs2288696; rs34503346; rs35316179; rs4733981; rs10958537; rs328301; rs6995988; rs113608876; rs35131262; rs7828172; rs60128210; rs1902171; rs12544187; rs34504029; bio/rs145713218; rs150878892; rs17182009; rs192636923; rs4593506; rs62504407; rs370681029; rs13274715; bio/rs77124900; rs72634766; rs7002593; rs6990092; bio/rs147255993; rs10088732; rs113120759; rs59960370; bio/rs17347923; rs35335280; rs77437987; rs72634730; rs7837897; rs76739044; rs138012741; rs17181995; rs61150883; rs4733948; rs12114449; bio/rs11785571; rs75923541; rs6994532; bio/rs12542436; rs75363590; rs11994860; rs144269447; rs73676485; rs17175708; rs12679830; rs34036147; rs11777965; rs72634773; rs12676965; rs4733883; rs2507712; rs13254512; rs74322494; rs7825208; rs7830301; bio/rs61746628; rs881301; rs876471; rs75203595; rs143196391; rs35935390; rs140333111; rs2593094; rs10106420; rs13275043; rs328293; rs11779106; rs77870750; rs72634752; bio/rs545291688; bio/rs182928359; bio/rs191625049; rs59281997; rs4739561; rs56247622; rs2924951; rs72634748; rs16887405; rs72634747; rs72634762; rs6474408; rs2445027; rs16887404; rs2445007; rs36105509; rs73674186; rs117993574; rs17175631; rs62503952; rs75533955; bio/rs151043937; rs112395365; rs6994556; rs78506403; rs72634756; rs17182023; rs76254545; rs12541651; rs16887675; rs10093843; rs6468435; rs58495890; rs117031182; rs11777840; rs1111460; rs55642511; rs72634767; rs3758102; bio/rs71552814; rs13264830; rs80204469; rs6468442; bio/rs113015423; rs881300; rs7388222; rs6474109; bio/rs117471926; bio/rs17362412; rs10958535; rs2924950; rs6992505; rs73674179; rs2067495; rs9650326; rs72636186; rs7821373; rs2445010; rs2461334; rs4733928; rs35422477; rs57103009; bio/rs17182254; bio/rs547252730; rs58771252; rs36042872; rs79891720; rs1900406; rs58191591; rs34544368; rs72630656; rs7829058; rs35804490; rs55916315; rs13278033; rs876472; rs28407048; rs4590405; rs7825829; bio/rs112643704; rs72634769; rs12677355; rs10090264; bio/rs76806580; rs56390495; rs4531014; rs73674184; rs62505495; rs7016216; rs35275553; rs10093754; rs6474108; rs34844754; rs7834231; bio/rs11787211; rs34873882; rs72634779; rs6474464; rs79195107; rs4308699; rs7012884; rs79131406; rs72634738; rs34813622; rs57613118; rs73674183; rs77509159; rs35638839; bio/rs28567491; rs328303; rs7014407; rs10808344; rs111449242; rs117089766",""
"bio/hg38_FGFR1","FGFR1","bio/rs117756469; rs11996010; rs7823851; rs7829871; rs13274384; rs13248612; rs72636191; rs7845393; rs6996321; rs2003265; rs4739558; rs2010445; rs3758101; rs11992272; rs13261164; rs35699533; rs4236755; rs16887681; rs10958704; rs62505473; rs72630687; rs7815496; rs4733980; rs145530646; rs72634742; rs111690554; rs7824271; rs7001431; rs328291; rs4342555; rs66504003; rs13279569; rs17182127; rs328294; rs141893102; rs142164645; rs930828; rs2445024; rs146505289; rs12114071; rs2568232; rs4601278; rs1541897; rs17175624; rs12677818; rs4647906; rs117787947; rs2568235; rs35305468; rs56307815; rs4292654; rs17182002; rs328300; rs881299; rs1447055; bio/rs35045158; rs4733979; rs17176009; rs4733946; rs73675815; rs200246954; rs113932779; rs3213849; rs28618161; rs7825758; rs4082204; rs4739560; rs72632529; rs113991129; rs28877647; rs36061954; rs183333881; rs17182134; rs2445020; rs2468964; rs6990295; rs79295570; rs2288696; rs34503346; rs35316179; rs4733981; rs10958537; rs328301; rs6995988; rs113608876; rs35131262; rs7828172; rs60128210; rs1902171; rs12544187; rs34504029; bio/rs145713218; rs150878892; rs17182009; rs192636923; rs4593506; rs62504407; rs370681029; rs13274715; bio/rs77124900; rs72634766; rs7002593; rs6990092; bio/rs147255993; rs10088732; rs113120759; rs59960370; bio/rs17347923; rs35335280; rs77437987; rs72634730; rs7837897; rs76739044; rs138012741; rs17181995; rs61150883; rs4733948; rs12114449; bio/rs11785571; rs75923541; rs6994532; bio/rs12542436; rs75363590; rs11994860; rs144269447; rs73676485; rs17175708; rs12679830; rs34036147; rs11777965; rs72634773; rs12676965; rs4733883; rs2507712; rs13254512; rs74322494; rs7825208; rs7830301; bio/rs61746628; rs881301; rs876471; rs75203595; rs143196391; rs35935390; rs140333111; rs2593094; rs10106420; rs13275043; rs328293; rs11779106; rs77870750; rs72634752; bio/rs545291688; bio/rs182928359; bio/rs191625049; rs59281997; rs4739561; rs56247622; rs2924951; rs72634748; rs16887405; rs72634747; rs72634762; rs6474408; rs2445027; rs16887404; rs2445007; rs36105509; rs73674186; rs117993574; rs17175631; rs62503952; rs75533955; bio/rs151043937; rs112395365; rs6994556; rs78506403; rs72634756; rs17182023; rs76254545; rs12541651; rs16887675; rs10093843; rs6468435; rs58495890; rs117031182; rs11777840; rs1111460; rs55642511; rs72634767; rs3758102; bio/rs71552814; rs13264830; rs80204469; rs6468442; bio/rs113015423; rs881300; rs7388222; rs6474109; bio/rs117471926; bio/rs17362412; rs10958535; rs2924950; rs6992505; rs73674179; rs2067495; rs9650326; rs72636186; rs7821373; rs2445010; rs2461334; rs4733928; rs35422477; rs57103009; bio/rs17182254; bio/rs547252730; rs58771252; rs36042872; rs79891720; rs1900406; rs58191591; rs34544368; rs72630656; rs7829058; rs35804490; rs55916315; rs13278033; rs876472; rs28407048; rs4590405; rs7825829; bio/rs112643704; rs72634769; rs12677355; rs10090264; bio/rs76806580; rs56390495; rs4531014; rs73674184; rs62505495; rs7016216; rs35275553; rs10093754; rs6474108; rs34844754; rs7834231; bio/rs11787211; rs34873882; rs72634779; rs6474464; rs79195107; rs4308699; rs7012884; rs79131406; rs72634738; rs34813622; rs57613118; rs73674183; rs77509159; rs35638839; bio/rs28567491; rs328303; rs7014407; rs10808344; rs111449242; rs117089766",""
"bio/mm10_Fgfr1","Fgfr1","",""
"bio/mm9_Fgfr1","Fgfr1","",""
```

chejennifer commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -142,13 +165,18 @@ const fetchData = async (
     // The rest of the promises are for column values
     const propResp = resp.slice(1);
     const values = {};
+    const sources = {};
```
It's slightly hard to follow when the same map holds provenanceIds as its values at the beginning of the function and then sourceUrls at the end. Maybe rename this to provenanceIds & then use sources later on.

dwnoble replied:
also add typing to the object

chejennifer commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -167,21 +196,55 @@ const fetchData = async (
+      for (const propertyExpr in sources[entity]) {
+        const sourceUrls = new Set<string>();
+        sources[entity][propertyExpr].forEach((provId) => {
+          sourceUrls.add(provIdToUrl[provId]);
```
we should filter out urls that are empty string

chejennifer commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -167,21 +196,55 @@ const fetchData = async (
+    props.columns.forEach((column) => {
+      const value = data.values[entity][column.propertyExpr].join("; ");
+      const source = data.sources[entity][column.propertyExpr].join("; ");
+      row.push(`${value},${source}`);
```
nit: might need to check that none of the values have a comma in them & maybe enclose the value in quotes if there is a comma

juliawu replied:
@dwnoble @chejennifer what do you think of matching the convention used in the datacommons client and encasing every string in quotes?

dwnoble replied:
That works, just make sure to escape the double quotes in that case: https://hawq.apache.org/docs/userguide/2.3.0.0-incubating/datamgmt/load/g-escaping-in-csv-formatted-files.html

chejennifer commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -167,21 +196,55 @@ const fetchData = async (
+    props.columns.forEach((column) => {
+      const value = data.values[entity][column.propertyExpr].join("; ");
+      const source = data.sources[entity][column.propertyExpr].join("; ");
+      row.push(`${value},${source}`);
```
I think you want to push value and source as separate items?

chejennifer commented on static/css/base.scss:
```diff
@@ -404,3 +404,8 @@ $headings-font-family: "Google Sans", $font-family-sans-serif;
   box-shadow: 0 0 0 0.2rem rgb(0 0 0 / 15%) !important;
 }
+
+/** Form Components */
```
I'm not sure what the best practice is for styled components, but I feel like for a component, we should have either have all its styling in a css file or all the styling in its .tsx file, not split over the two because that might make it hard to track down styles?

chejennifer reviewed (commented):
nice!

dwnoble commented on server/templates/explore.html:
```diff
@@ -25,6 +25,7 @@
  {% block head %}
  <link rel="stylesheet" href={{url_for('static', filename='css/explore.min.css', t=config['VERSION'])}} >
  <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
+ <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@20..48,100..700,0..1,-50..200" />
```
Looks like this adds a pretty large asset (network panel shows 3.1mb) - is there a way to grab fewer icons?

dwnoble commented on static/css/biomedical/biomedical_shared.scss:
```diff
@@ -41,3 +47,13 @@
   font-size: 10px;
 }
+
+.table {
```
nit: add a comment here describing what table this changes

dwnoble commented on static/js/components/form_components/buttons.tsx:
```diff
@@ -0,0 +1,127 @@
+  onClick: () => void;
+}
+
+function Button(props: ButtonProps): JSX.Element {
```
Nit: give a more specific name. IconButton ?

dwnoble commented on static/js/components/form_components/buttons.tsx:
```diff
@@ -0,0 +1,127 @@
+      className={`button ${props.class || ""}`}
+    >
+      {props.icon && (
+        <span className="material-symbols-outlined icon">{props.icon}</span>
```
When i load the page, the icon's name shows up as plain text for a split second before being replaced by the icon itself. is there any way to avoid showing the icon's plain text name?

dwnoble commented on static/js/components/form_components/buttons.tsx:
```diff
@@ -0,0 +1,127 @@
+export function CopyButton(props: CopyButtonProps): JSX.Element {
+  const [isClicked, setIsClicked] = useState<boolean>(false);
+
+  const onClick = () => {
```
Nit: Move to useEffect with clearTimeout https://felixgerschau.com/react-hooks-settimeout/

juliawu reviewed (commented):
Thanks for the reviews!

dwnoble commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -66,6 +70,10 @@ export function AnswerTableTile(props: AnswerTableTilePropType): JSX.Element {
         {props.title && <div className="answer-table-title">{props.title}</div>}
+        <DownloadButton
+          content={answerAsCsv}
+          filename={`${props.title}.csv`}
```
For the query "how many species of the flu are there?" I'm getting a file download named "undefined.csv"

dwnoble commented on static/js/components/form_components/icon_buttons.tsx:
```diff
@@ -0,0 +1,145 @@
+
+export function CopyButton(props: CopyButtonProps): JSX.Element {
+  const [isClicked, setIsClicked] = useState<boolean>(false);
+  const timerRef = useRef(null);
```
nit: add typing (`useRef<NodeJS.Timeout>` i think?)

dwnoble reviewed (commented):
Thanks Julia! Some style things: (cc @miss-o-soup):

Large text:
![94ZY93WrcDcZt97](https://github.com/datacommonsorg/website/assets/13766/3ca0e19c-d89e-4e92-8ca4-354189388096)

Bottom scroll bar:
![scroll1](https://github.com/datacommonsorg/website/assets/13766/552440b0-a4ba-4fd7-8e24-7813698b1c76)

Bottom scroll bar (continued)
![scroll2](https://github.com/datacommonsorg/website/assets/13766/334171e7-3cfd-4239-b7f3-9313095b2ad3)

juliawu commented:
> Thanks Julia! Some style things: (cc @miss-o-soup):


Per offline discussion, also made the table horizontally scrollable:

![Screenshot 2024-05-06 at 11 32 15 AM](https://github.com/datacommonsorg/website/assets/4034366/d5390ac1-93ba-4c7e-9d54-d5b78c0619df)

I've skipped making the first column sticky because I found it takes up too much real estate on moble.

chejennifer commented on static/css/biomedical/biomedical_shared.scss:
```diff
@@ -14,23 +14,24 @@
+  background-color: var(--main-content-background-color);
+}
+
+/* landing page needs uses a white background */
```
nit: needs uses -> either needs or uses?

juliawu replied:
fixed!

chejennifer commented on static/js/components/form_components/icon_buttons.tsx:
```diff
@@ -0,0 +1,145 @@
+    setIsClicked(true);
+  };
+
+  useEffect(() => {
```
nit: this isClicked + timer logic feels duplicated in the download and copy buttons, I wonder if it could be moved to IconButton?

juliawu replied:
Good callout, swapped to putting the logic in the IconButton

chejennifer commented on static/js/components/form_components/icon_buttons.tsx:
```diff
@@ -0,0 +1,145 @@
+    if (isClicked) {
+      timerRef.current = setTimeout(() => {
+        setIsClicked(false);
+      }, 2000);
```
nit: make 2000 a constant

juliawu replied:
Made a constant.

chejennifer commented on static/js/components/tiles/answer_message_tile.tsx:
```diff
@@ -40,43 +41,88 @@ export interface AnswerMessageTilePropType {
+      <b>
+        <i>{entityName}</i>
+      </b>
+      {title.slice(entityIndex + entityName.length, title.length)}
```
nit: for the slice method, you can leave the second argument empty to return the rest of the string, so this is equivalent to `title.slice(entityIndex + entityName.length)`

juliawu replied:
Thanks! did not know this. Simplified to not include the second argument.

chejennifer commented on static/js/components/tiles/answer_message_tile.tsx:
```diff
@@ -40,43 +41,88 @@ export interface AnswerMessageTilePropType {
+      paramsSerializer: stringifyFn,
+    });
+    const names = nameResp.data[props.entity] || [];
+    entityName = names ? names[0].value : "";
```
should check that names is not empty before accessing the first element?

chejennifer commented on static/js/components/tiles/answer_message_tile.tsx:
```diff
@@ -40,43 +41,88 @@ export interface AnswerMessageTilePropType {
-    return Promise.resolve(props.displayValue);
+): Promise<AnswerMessageTileData> => {
+  let entityName = "";
+  if (_.isEmpty(props.propertyExpr)) {
```
also do this if `props.displayValue` is true?

juliawu replied:
Added check for props.displayValue

chejennifer commented on static/js/components/tiles/answer_table_tile.tsx:
```diff
@@ -139,49 +163,121 @@ const fetchData = async (
           Array.from(entityResults);
+
+        // Get all sources associated with the values
+        const valSources = val.map((singleVal) => {
```
nit: call this valProvId instead so that we're consistent about provId vs source

juliawu replied:
Good call out, swapped to valProvId

chejennifer reviewed (commented):
thanks for the updates!

juliawu commented:
Updated per conversation with @miss-o-soup :

(1) First column is now sticky when scrolling

(2) Reduced font sizes in the table

New screenshot:
<img width="1010" alt="Screenshot 2024-05-07 at 11 45 46 AM" src="https://github.com/datacommonsorg/website/assets/4034366/8f964b2b-5281-474e-967f-0fa6ffe65521">

miss-o-soup commented:
Thank you! Julia, the other edits made to the mocks/table is to tone down the source information:
In the mocks it is set to medium 12 vs 16, and #5E5F59
Also the padding between the information and the source is too large, should be 2 to 4 px
See this example:
https://screenshot.googleplex.com/8nBFWrwgDNEKm6P.png

chejennifer reviewed (approved):
thanks for the changes!

juliawu commented:
@miss-o-soup source text was already at 12 medium size font, I added the #5E5F59 and changed the padding between the information and source to be 4px (2px looks very cramped on my screen)

New screenshot:
![Screenshot 2024-05-07 at 3 24 37 PM](https://github.com/datacommonsorg/website/assets/4034366/62e7a974-dbe5-4f68-8005-52530b898dc0)
